train_model.py

```diff
@@ -0,0 +1,199 @@
+"""This file implements the pipeline for training the model.
+
+This file defines a complete pipline to train to train and test
+a transformer-based model to predict intents and slots on the ATIS
+ans Snips datasets. Some of the code was in this file was taken 
+from https://www.tensorflow.org/tutorials/text/transformer.
+
+  Typical usage example:
+
+    python train_model.py --dataset='atis' 
+"""
+
+import tensorflow as tf
+import argparse
+import os
+import time
+from model import Net
+from util import load_data, evaluate, createVocabulary, loadVocabulary, create_masks
+
+class CustomSchedule(tf.keras.optimizers.schedules.LearningRateSchedule):
+  def __init__(self, d_model, warmup_steps=4000):
+    super(CustomSchedule, self).__init__()
+    
+    self.d_model = d_model
+    self.d_model = tf.cast(self.d_model, tf.float32)
+
+    self.warmup_steps = warmup_steps
+    
+  def __call__(self, step):
+    arg1 = tf.math.rsqrt(step)
+    arg2 = step * (self.warmup_steps ** -1.5)
+    
+    return tf.math.rsqrt(self.d_model) * tf.math.minimum(arg1, arg2)
+
+
+def loss_function(slot_real, intent_real, slot_pred, intent_pred, intent_loss_object, slot_loss_object):
+  mask = tf.math.logical_not(tf.math.equal(slot_real, 0))
+  slot_loss_ = slot_loss_object(slot_real, slot_pred)
+
+  mask = tf.cast(mask, dtype=slot_loss_.dtype)
+  slot_loss_ *= mask
+  slot_loss = tf.reduce_sum(slot_loss_)/tf.reduce_sum(mask)
+
+  intent_loss = intent_loss_object(intent_real, intent_pred)
+  intent_loss = tf.reduce_mean(intent_loss)
+
+  total_loss = slot_loss + intent_loss
+
+  return total_loss
+
+
+def main():
+
+  parser = argparse.ArgumentParser(allow_abbrev=False)
+  parser.add_argument("--batch_size", type=int, default=16, help="Batch size.")
+  parser.add_argument("--max_epochs", type=int, default=100, help="Max epochs to train.")
+  parser.add_argument("--num_layers", type=int, default=3, help="The number of transformer layers.")
+  parser.add_argument("--d_model", type=int, default=128, help="The dimensionality of the embeddings.")
+  parser.add_argument("--dff", type=int, default=512, help="The hidden layer size.")
+  parser.add_argument("--num_heads", type=int, default=8, help="The number of heads in the multi-headed attention layer.")
+  parser.add_argument("--dataset", type=str, default='atis', help="Type 'atis' or 'snips' ")
```

Comment:
Be consistent with your choice of string quote character within a file. Pick ' or " and stick with it. It is okay to use the other quote character on a string to avoid the need to \\ escape within the string.

Refer: http://google.github.io/styleguide/pyguide.html#310-strings